### PR TITLE
Task #1457217: Create FileFormatFactory for all file formats

### PR DESCRIPTION
**Pull request by assapir**

### CODE REVIEW

File: sdk/file_sdk/Common/file_format_factory.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+namespace {
+static const vector<string> kXMPExtentions {".jpg", ".jpeg", ".jpe", ".jif", ".jfif", ".jfi",".png", ".tif", ".tiff", ".gif", ".dng", ".psd"};
+static const vector<string> kCompoundFileExtentions {".doc", ".doc", ".xls", "xlt", ".ppt", ".pps", ".pot", ".vsd", ".vdw", ".vst", ".pub", ".sldprt", ".slddrw", ".sldasm", ".mpp", ".mpt"};
+static const vector<string> kOPCFileExtentions {".docx", ".docm", ".dotm", ".dotx", ".xlsx", ".xltx", ".xltm", ".xlsm", ".xlsb", ".pptx", ".ppsx", ".pptm", ".ppsm", ".potx", ".potm",  ".vsdx", ".vsdm", ".vssx", ".vssm", ".vstx", ".vstm", ".xps", ".oxps", ".dwfx"};
+static const vector<string> kPDFExtentions {".pdf"};
+static const vector<string> kPfileExtentions {".pfile", ".ppdf", ".ptxt", ".pxml", ".pjpg", ".pjpeg", ".ppng", ".ptif", ".ptiff", ".pbmp", ".pgif", ".pjpe", ".pjfif", ".pjt"};
+
+bool in_array(const string &value, const vector<string> &array) {
+    return std::find(array.begin(), array.end(), value) != array.end();
```
Comment (shenhavba) [inline sdk/file_sdk/Common/file_format_factory.cpp]: >   [](start = 2, length = 2)

2 spaces.
and i don't really care...

File: sdk/file_sdk/Common/file_format_factory.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+FileFormatFactory::FileFormatFactory()
+{
+}
+
+std::shared_ptr<IFileFormat> FileFormatFactory::Create(std::shared_ptr<IStream> fileStream, const string& extension) {
+
+  if (extension.empty())
+  {
+    // TODO: add log
+    throw rmscore::exceptions::RMSInvalidArgumentException("The extention is empty");
```
Comment (shenhavba) [inline sdk/file_sdk/Common/file_format_factory.cpp]: why do you throw RMSInvalidArgumentException? its not related to rms. i think there is built in exception for invalid argument #Resolved

File: sdk/file_sdk/Common/file_format_factory.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+namespace file {
+
+FileFormatFactory::FileFormatFactory()
+{
+}
+
+std::shared_ptr<IFileFormat> FileFormatFactory::Create(std::shared_ptr<IStream> fileStream, const string& extension) {
+
+  if (extension.empty())
+  {
```
Comment (shenhavba) [inline sdk/file_sdk/Common/file_format_factory.cpp]: > { [](start = 1, length = 2)

should be in the same line of the if in all places #Resolved

File: sdk/file_sdk/Common/Common.pro [hunk 4]
```diff
@@ -2,23 +2,27 @@ REPO_ROOT = $$PWD/../../..
 DESTDIR   = $$REPO_ROOT/bin/filesdk
 TARGET    = modfilecommon
 
-
 TEMPLATE  = lib
 CONFIG += staticlib warn_on c++11 debug_and_release
-INCLUDEPATH += $$REPO_ROOT/sdk/rmscrypto_sdk/CryptoAPI
+INCLUDEPATH += $$REPO_ROOT/sdk/rmscrypto_sdk/CryptoAPI $$REPO_ROOT/sdk/rms_sdk/ModernAPI $$REPO_ROOT/sdk/file_sdk
```
Comment (vakuras) [inline sdk/file_sdk/Common/Common.pro]: why do you need the file_sdk? you are in the same project. or not? #WontFix

File: sdk/file_sdk/Common/file_format.h [hunk 5]
```diff
@@ -23,7 +23,7 @@ class FileFormat : public IFileFormat
   FileFormat(shared_ptr<IStream> file, const string& extension);
 
   // Reads the tags from the file
-  virtual const vector<Tag>& ReadTags() = 0;
+  virtual const vector<Tag> ReadTags() = 0;
```
Comment (vakuras) [inline sdk/file_sdk/Common/file_format.h]: why did you remove the &? #Resolved

File: sdk/file_sdk/Common/file_format_factory.cpp [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+using mip::file::XMPFileFormat;
+
+namespace {
+static const vector<string> kXMPExtentions {".jpg", ".jpeg", ".jpe", ".jif", ".jfif", ".jfi",".png", ".tif", ".tiff", ".gif", ".dng", ".psd"};
+static const vector<string> kCompoundFileExtentions {".doc", ".doc", ".xls", "xlt", ".ppt", ".pps", ".pot", ".vsd", ".vdw", ".vst", ".pub", ".sldprt", ".slddrw", ".sldasm", ".mpp", ".mpt"};
+static const vector<string> kOPCFileExtentions {".docx", ".docm", ".dotm", ".dotx", ".xlsx", ".xltx", ".xltm", ".xlsm", ".xlsb", ".pptx", ".ppsx", ".pptm", ".ppsm", ".potx", ".potm",  ".vsdx", ".vsdm", ".vssx", ".vssm", ".vstx", ".vstm", ".xps", ".oxps", ".dwfx"};
+static const vector<string> kPDFExtentions {".pdf"};
+static const vector<string> kPfileExtentions {".pfile", ".ppdf", ".ptxt", ".pxml", ".pjpg", ".pjpeg", ".ppng", ".ptif", ".ptiff", ".pbmp", ".pgif", ".pjpe", ".pjfif", ".pjt"};
+
+bool in_array(const string &value, const vector<string> &array) {
```
Comment (vakuras) [inline sdk/file_sdk/Common/file_format_factory.cpp]: this should be a private method of FileFormatFactory #Resolved

File: sdk/file_sdk/XMPFileFormat/xmp_file_format.cpp [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+namespace mip {
+namespace file {
+
+XMPFileFormat::XMPFileFormat(shared_ptr<IStream> file, const string& extension)
+  : FileFormat(file, extension)
+{
+}
+
+const vector<Tag> XMPFileFormat::ReadTags() {
+  return vector<Tag>();
```
Comment (vakuras) [inline sdk/file_sdk/XMPFileFormat/xmp_file_format.cpp]: should throw a not implemented exception for now #Resolved

File: sdk/file_sdk/PFileFileFormat/pfile_file_format.cpp [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+namespace mip {
+namespace file {
+
+PFileFileFormat::PFileFileFormat(shared_ptr<IStream> file, const string& extension)
+  : FileFormat(file, extension)
+{
+}
+
+const vector<Tag> PFileFileFormat::ReadTags() {
+  vector<Tag> tmp;
```
Comment (vakuras) [inline sdk/file_sdk/PFileFileFormat/pfile_file_format.cpp]: should throw a not implemented exception for now #Resolved

File: sdk/file_sdk/PDFFileFormat/pdf_file_format.cpp [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+namespace mip {
+namespace file {
+
+PDFFileFormat::PDFFileFormat(shared_ptr<IStream> file, const string& extension)
+  : FileFormat(file, extension)
+{
+}
+
+const vector<Tag> PDFFileFormat::ReadTags() {
+  return vector<Tag>();
```
Comment (vakuras) [inline sdk/file_sdk/PDFFileFormat/pdf_file_format.cpp]: should throw a not implemented exception for now #Resolved

**Review by vakuras — CHANGES_REQUESTED** [review]

File: sdk/file_sdk/Common/Common.pro (hunk 4 above)
Comment (assapir) [reply sdk/file_sdk/Common/Common.pro]: we should not use .. in #include

---
In reply to: [134661416](https://github.com/AzureAD/rms-sdk-for-cpp/pull/134#discussion_r134661416) [](ancestors = 134661416)

**Review by shenhavba — APPROVED** [review]
:shipit:

File: sdk/file_sdk/Common/file_format_factory.cpp [hunk 10]
```diff
@@ -38,39 +33,37 @@ FileFormatFactory::FileFormatFactory()
...
-std::shared_ptr<IFileFormat> FileFormatFactory::Create(std::shared_ptr<IStream> fileStream, const string& extension) {
+// static
+bool FileFormatFactory::in_array(const std::string &value, const vector<std::string> &array) {
+    return std::find(array.begin(), array.end(), value) != array.end();
+}
 
-  if (extension.empty())
-  {
+// static
+std::shared_ptr<IFileFormat> FileFormatFactory::Create(std::shared_ptr<IStream> fileStream, const string& extension) {
```
Comment (vakuras) [inline sdk/file_sdk/Common/file_format_factory.cpp]: just stream

**Review by vakuras — APPROVED** [review]

**Pull request merged**